doc/sphinx/proof-engine/ltac.rst

```diff
@@ -215,51 +202,42 @@ A sequence is an expression of the following form:
    :n:`v__2` is applied to all the goals produced by the prior
    application. Sequence is associative.
 
+   .. todo PR I don't see the distinction between evaluating an ltac expression
+      and applying it--how are they not the same thing?
```

Comment:
Still seems too imprecise.  Is the `match` considered a tactic?  If so, what other Ltac constructs are handled in evaluation rather than application.  Does "tactic" mean simple tactic or a tactic expression?

The language in the "Semantics" section makes an undefined distinction between evaluation and application:

> Tactic expressions can only be applied in the context of a proof. The
evaluation yields either a term, an integer or a tactic.

Reply:
> Does "tactic" mean simple tactic or a tactic expression?

I guess it means something in between. Yes we should clarify all of this. But that's definitely out of scope of a PR trying to fix the syntax. In further PRs trying to explain things better, I would involve more people that are experts, but it would be smaller PRs that would be less hard to follow.

Reply:
OK, punt

Reply:
> If so, what other Ltac constructs are handled in evaluation rather than application.

I've explained this on coq-club and in the issue tracker a couple times, and I think some others have explained it as well; here are some relevant posts:
I explain it at length in https://github.com/coq/coq/issues/4732#issuecomment-337542429 (in particular, this explains what happens during tactic expression evaluation time)
See also:
https://sympa.inria.fr/sympa/arc/coq-club/2016-10/msg00057.html
https://sympa.inria.fr/sympa/arc/coq-club/2018-04/msg00092.html
https://github.com/coq/coq/issues/5235#issuecomment-337549681
http://coqhott.gforge.inria.fr/blog/coq-tactic-engine/
@aspiwack 's original explanation is at https://sympa.inria.fr/sympa/arc/coqdev/2014-12/msg00130.html and is quite excellent.

While I'm bringing up ancient history, I wonder if @aspiwack can confirm my guess; in [your email](https://sympa.inria.fr/sympa/arc/coqdev/2014-12/msg00130.html), you say:

> Two of these have been mentioned on this thread: [match] doesn't give the strange "Immediate match producing tactics not allowed in local definitions." error. Instead, it evaluates eagerly like in a cbv language. And [lazymatch] is now more uniform with [match] as it also evaluates its branches eagerly.
> 
> The latter change I've made because: well, first and foremost, I had always thought it was the semantics, so when I had to modify the implementation of lazymatch to use the new tactic API, I just did it that way. In the process of doing so, I discovered how [lazymatch] worked before (see below) in order to try and be lazy and was a tad horrified. It didn't work. Now, there may be a way to make [lazymatch] lazy without these issue, but working ones way around Ltac's interpreter is really hard, and I simply don't know how to do it without rewriting it all (quite a perilous endeavour). Plus, I'm not even sure it is a good idea, I feel it more predictable when the various variants of match behave as uniformly as possible. I may be wrong of course.

You didn't explain why you removed the "Immediate match producing tactics not allowed in local definitions." error. I hypothesized [in this email](https://sympa.inria.fr/sympa/arc/coq-club/2018-04/msg00092.html):

> Historical fact: in Coq <= 8.4, it gave an error message.  "Immediate match producing tactics not allowed in this location" or something.  This error message was removed in Coq 8.5, when Ltac switched to the new proof engine.
> 
> My hypothesis is that the original author of Ltac wanted a language that could both run tactics and produce terms, implemented it in the first way they thought of, noticed that this sort of thing was weird and unintuitive, and added an error message forbidding it.  This quirk of implementation, that `match` runs its branches doing tactic _expression_ evaluation, was nearly unobservable (you could only notice it by nesting match in lazymatch and noticing that you could get an unexpected failure level).
> Many years later, when the devs rewrote the tactic engine to be saner, they had no collective recollection of why Ltac was the way it was (and, after all, this is all merely my guessing based on the feel of Ltac as a language), and saw no reason to keep the error message, so they permitted this behavior.  They could not change the fact that match runs early, though, because that would break almost every tactic that constructs a term.

I'm curious @aspiwack if you recall whether or not my hypothesis is correct.

Reply:
> the distinction between evaluating an ltac expression and applying it

Btw, I think "applying" is a misleading description of the thing that is not evaluation.  I would call it "running" or "execution" of the tactic.  (Tactic expression evaluation already has a notion of application which is not this one, namely, I can write `let f _ := idtac in f ()` and `f ()` is an application of `f` to `()` which happens during tactic expression evaluation, whereas `idtac` is "applied" to the current goal during execution/runtime.)

Reply:
Thanks for the background--I put a link in the rst file as a todo.

BTW, do you happen to know how multimatch differs from match and lazymatch?  I couldn't find an example, let alone explain.

Reply:
> I'm curious @aspiwack if you recall whether or not my hypothesis is correct.

I can't say that I fully remember all of this to be honest. But I think that the eager match was simply not possible in the old tactic model. Essentially you had to build a tree of tactic and execute it in one step. Or maybe it would have been possible, but contrived.

The difficulty with Ltac, was that it was rather difficult to fit in an ML-like semantics with the tools of the Milner-style tactic engine. But the new tactic engine was simply a monad. Monads are almost just another word for an ML-like semantics. So suddenly some operations which were barely thinkable became natural.

And I assume that this is what motivated this particular change. Does it make sense?